File: src.csharp/AlphaTab/Core/Lazy.cs
namespace AlphaTab.Core
{
    public class Lazy<T> where T:class?
    {
        private readonly System.Func<T> _factory;

        private bool _created = false;
        private T _value;

        public Lazy(System.Func<T> factory)
        {
            _factory = factory;
            _value = null!;
        }

        public T Value
        {
            get
            {
                if (!_created)
                {
                    _value = _factory();
                }
                return _value;
            }
        }
    }
}

File: src.kotlin/alphaTab/src/commonMain/kotlin/alphaTab/core/Lazy.kt
package alphaTab.core

class Lazy<T : Any?> {
    private val _factory: () -> T
    private var _created:Boolean = false;
    private var _value:T;

    public constructor(factory:() -> T) {
        this._factory = factory;
    }

    public val value:T
        get() {
            if(!_created) {
                _value = _factory()
            }
            return _value;
        }
}
